### PR TITLE
Add ffi_integration_test_{android,ios} which runs dart-lang/sdk's positive `dart:ffi` tests

### PR DESCRIPTION
**Pull request by mkustermann**
This change adds ffi_integration_test_{android,ios} tests, which an be
launched via:

    dev/devicelab % pub get
    dev/devicelab % dart bin/run.dart -t ffi_integration_test_android
    dev/devicelab % dart bin/run.dart -t ffi_integration_test_ios

The change includes tests extracted from dart-lang/sdk's "ffi" test
suite. The corresponding files are checked into:

  * dev/integration_tests/ffi/ios/Classes/*
  * dev/integration_tests/ffi/lib/src/generated/*

Since those tests are not hand written in flutter/flutter, they don't
correspond to flutter's style guide. We therefore exclude running the
analyzer/linter on via `analysis_options.yaml`.

### CODE REVIEW

**Comment by dcharkes** [comment]
@mkustermann Do I need some kind of setup step somewhere?

```
Unhandled exception:
JSON-RPC error -32000: Server error
package:json_rpc_2/src/client.dart 110:64              Client.sendRequest
package:json_rpc_2/src/peer.dart 79:15                 Peer.sendRequest
package:vm_service_client/src/scope.dart 64:23         Scope.sendRequestRaw
package:vm_service_client/src/isolate.dart 361:19      VMIsolateRef.invokeExtension
package:flutter_devicelab/framework/runner.dart 72:59  runTask
===== asynchronous gap ===========================
bin/run.dart 64:47                                     main
dart:isolate-patch/isolate_patch.dart 305:32           _startIsolate.<fn>
dart:isolate-patch/isolate_patch.dart 174:12           _RawReceivePortImpl._handleMessage
```

**Review by dcharkes — APPROVED** [review]
I'm not able to run the devicelab command, but I am able to run the flutter app that runs the test:

```
dev/integration_tests/ffi $ flutter run -d all
```

So from an dart:ffi dev point of view, I can confirm that the tests work.

Please request review from a Flutter dev as well.

**Comment by goderbauer** [comment]
Cirrus seems unhappy with this change. Can you take a look at that?

**Comment by goderbauer** [comment]
Why do we need to have a copy of the ffi tests in the flutter repository?

**Comment by mkustermann** [comment]
> Why do we need to have a copy of the ffi tests in the flutter repository?

For two reasons:

- Dart's CI has right now almost no coverage on actual end user devices (no iOS device tests, only a single Android device type). We have encountered e.g. iOS-specific bugs before (uncovered with this PR actually) and there's even some open bugs (like https://github.com/flutter/flutter/issues/54948). Flutter's device lab is much more broad and allows us to test on more representative set of devices. (**)
- We want to test `dart:ffi` inside an actual flutter app (i.e. with the flutter embedder) and not just in combination with our standalone `dart` embedder, as we do on Dart's CI.

I agree that this is not an ideal solution. The best solution would be if we had CI support for checking out multiple git repositories generating this integration test flutter app from the tests in the Dart sdk and run the test.

We actually have a Head-Head-Head setup (for flutter/flutter, flutter/engine, dart-lang/sdk), but there's no devices we can run it on atm.

(**) We have asked for getting devices and get them integrated into our CI many months ago, but it was blocked for a long time (@athomas might know more).

**Comment by goderbauer** [comment]
Testing ffi on actual devices sounds great. How's the duplicated code kept in sync between the repositories? Would it maybe be better to just have the devicelab test checkout the appropriate code (at a predefined revision) and run it (cc @godofredoc)? 

I don't think it is ideal to have this copy in the flutter repository (also considering that it doesn't follow Flutter's coding conventions).

File: dev/integration_tests/ffi/lib/src/generated/aliasing_test.dart [hunk 1]
```diff
@@ -0,0 +1,216 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
```
Comment (Hixie) [inline dev/integration_tests/ffi/lib/src/generated/aliasing_test.dart:5]: This repo must only have code with the "Copyright 2014 The Flutter Authors" copyright.

Comment (mkustermann) [reply dev/integration_tests/ffi/lib/src/generated/aliasing_test.dart:5]: @Hixie Can I use flutter-only copyright even though the dart files's source-of-truth is under dart copyright? If so, I'd be happy to do it.

**Comment by mkustermann** [comment]
> Testing ffi on actual devices sounds great. How's the duplicated code kept in sync between the repositories? Would it maybe 
> be better to just have the devicelab test checkout the appropriate code (at a predefined revision) and run
> it (cc @godofredoc)?

My original intention was to update the ffi integration tests regularly by importing the of the source-of-truth from dart-lang/sdk.

Theoretically, I agree it would be nice if the devicelab test can get the sources itself, build the app and run it. Practically I think pulling it from source of truth (dart-lang/sdk) would be very problematic - (our dart-lang/sdk is multi-GB). We could export the files from dart-lang/sdk to a GCS bucket and make the devicelab test pull it from there. Though IMHO even this approach is rather complex and hard to maintain.

As might have been obvious from the very slow response rate (apologies for that!), I don't have too many cycles to spend on this. If you prefer to have a more complicated solution than the current PR, I would ask @dcharkes (our main ffi person) to drive it.

As we are going towards a FFI 1.0 in the second half of the year, we really want to have a wider device coverage for our ffi tests.

> This repo must only have code with the "Copyright 2014 The Flutter Authors" copyright.

(Ideally the Dart team's head-head-head builder would have access to all devices in the lab. Then we could setup this ffi integration test on our side instead of flutter/flutter. But it doesn't seem like this is achievable for the foreseeable future.)

**Comment by dnfield** [comment]
I'm going to close this to get it off of review queues - if we plan to come back to this, we can re-open (but based on the comments in the linked issue it seems like it's not planned for right now).

**Pull request closed without merging**